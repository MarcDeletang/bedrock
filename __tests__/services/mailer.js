const importApp = require('../../app');
const importMock = require('../../app/mocks/mailer.ts');
const importBase = require('../base.js');
const { MailerService } = importApp;
const { mockMandrill } = importMock;
const { rejectGenerator, handleResolveSnap, handleRejectSnap } = importBase;

const apiKey = 'foo';
const address = 'bar';
const name = 'baz';

const email = 'foo@bar.baz';
const firstName = 'foo';
const lastName = 'baz';

const html = '';
const subject = '';
const from_email = email;
const from_name = email;

const config = {
  mailer: {
    apiKey,
    address,
    name,
  },
};

const recipient = {
  email,
  firstName,
  lastName,
};

const failInit = rejectGenerator(MailerService.init);

describe('Testing the mailer service initialisation', () => {
  it('Check the service shape', () => {
    expect(MailerService).toMatchSnapshot();
  });

  it('Reject with error (config.mailer)', failInit({}, mockMandrill));

  it('Reject with error (config.mailer.apiKey)', failInit({ mailer: {} }, mockMandrill));

  it('Reject with error (config.mailer.address)', failInit({ mailer: { apiKey } }, mockMandrill));

  it('Reject with error (config.mailer.name)', failInit({ mailer: { apiKey, address } }, mockMandrill));

  it('Load in prod without error', () => {
    return handleResolveSnap(MailerService.init, config, mockMandrill, true);
  });

  it('Load without error', () => {
    return handleResolveSnap(MailerService.init, config, mockMandrill);
  });

});

const failSendMail = rejectGenerator(MailerService.sendMail);

describe('Testing the mailer', () => {
  it('Send mail no error', () => {
    return handleResolveSnap(MailerService.sendMail, 'subject', 'content', recipient);
  });

  it('Send mails no error', () => {
    return handleResolveSnap(MailerService.sendMails, 'subject', 'content', [recipient]);
  });

  it('Send mails error recipient', () => {
    return handleRejectSnap(MailerService.sendMails, 42, 'content', null);
  });

  it('Send mail error recipient', failSendMail(42, 'content', null));

  it('Send mail error recipient email', failSendMail(42, 'content', {}));

  it('Send mail error recipient firstName', failSendMail(42, 'content', { email }));

  it('Send mail error recipient lastName', failSendMail(42, 'content', { email, firstName }));

  it('Send mail error subject', failSendMail(42, 'content', recipient));

});
