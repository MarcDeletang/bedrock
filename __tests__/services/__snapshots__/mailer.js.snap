// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the mailer Send mail error recipient 1`] = `[Error: recipient]`;

exports[`Testing the mailer Send mail error recipient email 1`] = `[Error: recipient.email]`;

exports[`Testing the mailer Send mail error recipient firstName 1`] = `[Error: recipient.firstName]`;

exports[`Testing the mailer Send mail error recipient lastName 1`] = `[Error: recipient.lastName]`;

exports[`Testing the mailer Send mail error subject 1`] = `
Object {
  "err": "Invalid subject",
  "message": 42,
}
`;

exports[`Testing the mailer Send mail no error 1`] = `
Object {
  "async": true,
  "ip_pool": "Main Pool",
  "message": Object {
    "from_email": "bar",
    "from_name": "baz",
    "html": "content",
    "subject": "subject",
    "to": Array [
      Object {
        "email": "foo@bar.baz",
        "name": "foo baz",
        "type": "to",
      },
    ],
  },
  "send_at": "2000-01-01 00:00:00",
}
`;

exports[`Testing the mailer Send mails error recipient 1`] = `[TypeError: Cannot read property 'forEach' of null]`;

exports[`Testing the mailer Send mails no error 1`] = `
Object {
  "async": true,
  "ip_pool": "Main Pool",
  "message": Object {
    "from_email": "bar",
    "from_name": "baz",
    "html": "content",
    "subject": "subject",
    "to": Array [
      Object {
        "email": "foo@bar.baz",
        "name": "foo baz",
        "type": "to",
      },
    ],
  },
  "send_at": "2000-01-01 00:00:00",
}
`;

exports[`Testing the mailer service initialisation Check the service shape 1`] = `
Object {
  "init": [Function],
  "sendMail": [Function],
  "sendMails": [Function],
}
`;

exports[`Testing the mailer service initialisation Load in prod without error 1`] = `
Object {
  "address": "bar",
  "client": Mandrill {
    "apikey": "foo",
    "debug": false,
    "exports": Exports {
      "master": [Circular],
    },
    "inbound": Inbound {
      "master": [Circular],
    },
    "internal": Internal {
      "master": [Circular],
    },
    "ips": Ips {
      "master": [Circular],
    },
    "messages": Messages {
      "master": [Circular],
    },
    "metadata": Metadata {
      "master": [Circular],
    },
    "rejects": Rejects {
      "master": [Circular],
    },
    "senders": Senders {
      "master": [Circular],
    },
    "subaccounts": Subaccounts {
      "master": [Circular],
    },
    "tags": Tags {
      "master": [Circular],
    },
    "templates": Templates {
      "master": [Circular],
    },
    "urls": Urls {
      "master": [Circular],
    },
    "users": Users {
      "master": [Circular],
    },
    "webhooks": Webhooks {
      "master": [Circular],
    },
    "whitelists": Whitelists {
      "master": [Circular],
    },
  },
  "name": "baz",
}
`;

exports[`Testing the mailer service initialisation Load without error 1`] = `
Object {
  "address": "bar",
  "client": Object {
    "address": "foo",
    "messages": Object {
      "send": [Function],
    },
    "name": "bar",
  },
  "name": "baz",
}
`;

exports[`Testing the mailer service initialisation Reject with error (config.mailer) 1`] = `[Error: config.mailer]`;

exports[`Testing the mailer service initialisation Reject with error (config.mailer.address) 1`] = `[Error: config.mailer.address]`;

exports[`Testing the mailer service initialisation Reject with error (config.mailer.apiKey) 1`] = `[Error: config.mailer.apiKey]`;

exports[`Testing the mailer service initialisation Reject with error (config.mailer.name) 1`] = `[Error: config.mailer.name]`;
